Fullcube/validates uniqueness of case insensitive

beeman opened the pull request:
This is an improved version of this PR https://github.com/strongloop/loopback-datasource-juggler/pull/798

---

This patch adds an option to check the uniqueness of a value in a case insensitive way.

Should fix #607, though I'm not sure if I overlooked any edge cases.

Amir-61 commented:
Thanks @beeman for the PR. There are some Merge conflicts, could you please rebase your code on top of `master` and squash your commit into a single commit following [this standard](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html). Thanks!

mrfelton commented:
@Amir-61 I've rebased this patch off master.

Amir-61 commented:
@mrfelton @beeman the commit message is not following the standards defined [here](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html); thats why the test failed. Could you please fix it.

Amir-61 commented:
you can use `git commit --amend` to change the commit.

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

mrfelton commented:
@Amir-61 I've adjusted the commit message.

Amir-61 commented:
@slnode test please

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

Amir-61 commented:
This feature and the logic make sense to me; I added `needs-doc` label as well.

@beeman/ @mrfelton Can we please have specific test for conf `{ ignoreCase: false }` as well? 

@bajtos What is your thought about this feature?

Amir-61 commented:
@beeman @mrfelton Thanks for contribution 👍

Amir-61 commented on test/validations.test.js:
```diff
@@ -525,6 +525,52 @@ describe('validations', function() {
+        });
+      })).should.be.false;
+    });
+
```
Nitpick: Please remove this extra blank line

Amir-61 commented on test/validations.test.js:
```diff
@@ -525,6 +525,52 @@ describe('validations', function() {
       });
     });
+
+    it('should validate uniqueness case insensitive', function(done) {
```
Please do not start test names with `should`, even though we have followed some bad practices in our previous code. Please see our [style-guide](https://github.com/strongloop/loopback-contributor-docs/blob/master/style-guide.md#test-naming)

Please fix it for other tests as well.

bajtos commented:
@Amir-61 I think the feature request to have a case-insensitive unique constraint is a valid one.

The only concern I have is whether we have an operator that can be used in all connectors we support (SQL and NoSQL).

I'll leave further review to you @Amir-61 and @raymondfeng for final approval if needed.

beeman commented:
Any updates on this PR? :)

Amir-61 commented:
>The only concern I have is whether we have an operator that can be used in all connectors we support (SQL and NoSQL).

Thats my concern as well... I'm not sure if all supported connectors have this functionality... @raymondfeng any thought on this?

Amir-61 commented:
@slnode test please

beeman commented:
Hey @Amir-61, thanks for your reply. 

Can I do anything to make sure this works? Would it work to test against MySQL/Oracle manually or are there better ways to do this?

Amir-61 commented:
>Can I do anything to make sure this works? Would it work to test against MySQL/Oracle manually or are there better ways to do this?

AFICT tests in `test/validations.test.js` are not run by connectors, whereas for example tests in `test\persistentence-hook.suite.js` are run by each connectors (dependencies); hence connectors which are not running your added tests in `test/validations.test.js`. I believe the only way is to add test for each connector individually to make sure they are passing. @raymondfeng  @bajtos for more thought?

Amir-61 commented:
@slnode test please

beeman commented:
@Amir-61 I've rebased the commit to the latest master, can you please consider merging or tell me what needs to be done to get it in? Thanks!

Amir-61 commented:
@beeman Thanks for the update. @loay is going to take care of that.

loay commented:
Hi @beeman There are few linting errors in this pr. 
Please run this command `./node_modules/.bin/eslint --fix .` as it should fix these issues, and/or will tell you what you need to do to get it fixed. 

Also, please make sure you have addressed Amir previous comments. 
We will give it another round of review so we can merge it asap
Thanks for your contribution.

loay commented:
@slnode test please

beeman commented:
@Amir-61 @loay thanks!

I've fixed the linting errors and updated the test names based on the previous comments.

loay commented:
@slnode test please

bajtos commented on test/validations.test.js:
```diff
@@ -517,6 +517,51 @@ describe('validations', function() {
+      var u = new User({email: 'me+me@my.com'});
+      Boolean(u.isValid(function(valid) {
+        valid.should.be.true;
+        u.save(function() {
```
`u.save` can fail with an error, please handle it.

```
u.save(function(err) {
  if (err) return done(err);
  // ...
});
```

bajtos commented on test/validations.test.js:
```diff
@@ -517,6 +517,51 @@ describe('validations', function() {
+      var u = new User({email: 'also+me@my.com'});
+      Boolean(u.isValid(function(valid) {
+        valid.should.be.true;
+        u.save(function() {
```
`u.save` can fail with an error, please handle it.

bajtos commented on test/validations.test.js:
```diff
@@ -517,6 +517,51 @@ describe('validations', function() {
+      User.validatesUniquenessOf('email', {ignoreCase: true});
+      var u = new User({email: 'me+me@my.com'});
+      Boolean(u.isValid(function(valid) {
+        valid.should.be.true;
```
https://github.com/moll/js-must#beware-of-libraries-that-assert-on-property-access

I believe `.true` is a function in recent versions of shouldjs.

```js
valid.should.be.true();
```

bajtos commented on test/validations.test.js:
```diff
@@ -517,6 +517,51 @@ describe('validations', function() {
+      var u = new User({email: 'hey'});
+      Boolean(u.isValid(function(valid) {
+        valid.should.be.true;
+        u.save(function() {
```
`u.save` can fail with an error, please handle it.

loay commented:
@bajtos I was also thinking of having at least one test with `ignoreCase: false` to validate the behavior. What do you think?

beeman commented:
@bajtos thanks for reviewing, I've addressed your comments.

@loay good point, I've added all 3 tests with `ignoreCase` set to `false`.

Please let me know if there's anything else to do to land this.

loay commented:
@slnode test please

loay commented:
@beeman thanks .. let's run the test one more time to ensure everything is fine on CI before merging

loay commented:
@beeman can you please rebase? Thanks.

beeman commented:
@loay Rebase done 👍

loay commented:
@slnode test please

bajtos commented:
As I was re-reading the discussion, https://github.com/strongloop/loopback-datasource-juggler/pull/818#issuecomment-256737192 caught my attention.

> > Can I do anything to make sure this works? Would it work to test against MySQL/Oracle manually or are there better ways to do this?

> AFICT tests in `test/validations.test.js` are not run by connectors, whereas for example tests in `test\persistentence-hook.suite.js` are run by each connectors (dependencies); hence connectors are not running your added tests in `test/validations.test.js`. I believe the only way to verify is to add tests for each connector individually to make sure they are passing. @raymondfeng  @bajtos for more thought?

I think most connectors are running `test/common.batch.js`. It should be enough to add `require('test/validations.test.js')` to that file in order to start running validation tests in each connector. However, I think this should be done in a standalone pull request, because we may need to fix some of the validation tests to avoid assumptions that don't hold for all downstream connectors.

loay commented:
@beeman Thanks for your contribution.

beeman commented:
Thanks for your input and help getting it landed guys 👍🏼
